File: Linked Lists/removeDups.js
/**
 * Solution 1 - Set track
 *
 * Complexity Analysis
 * O(n) time | O(n) space
 *
 * Remove duplicates from linked list
 *
 * @param {object} linkedList
 * @returns {object} Linked list with no duplicates
 */
function removeDups(linkedList = {}) {
    // Gracefully handle type and Falsy values
    if (typeof linkedList !== 'object' || Object.keys(linkedList || {}).length === 0) {
        console.error('Parameter should be a valid non-empty object');
        return false;
    }

    const trackedData = new Set();

    let prevNode;
    let currNode = linkedList; // head

    // Traverse linked list till its tail
    while (currNode !== null) {
        if (trackedData.has(currNode.data)) {
            prevNode.next = currNode.next;
        } else {
            trackedData.add(currNode.data);
            prevNode = currNode;
        }

        currNode = currNode.next;
    }

    return linkedList;
}

/**
 * Solution 2 - Inner loop
 *
 * Complexity Analysis
 * O(n^2) time | O(1) space
 *
 * Remove duplicates from linked list
 *
 * @param {object} linkedList
 * @returns {object} Linked list with no duplicates
 */
function removeDups(linkedList = {}) {
    // Gracefully handle type and Falsy values
    if (typeof linkedList !== 'object' || Object.keys(linkedList || {}).length === 0) {
        console.error('Parameter should be a valid non-empty object');
        return false;
    }

    let currNode = linkedList; // head

    // Traverse linked list till its tail
    while (currNode !== null) {
        let prevInnerNode = currNode;
        let innerNode = currNode.next;

        while (innerNode !== null) {
            if (currNode.data === innerNode.data) {
                prevInnerNode.next = innerNode.next;
            }

            prevInnerNode = innerNode;
            innerNode = innerNode.next;
        }

        currNode = currNode.next;
    }

    return linkedList;
}

const testCases = [
    // Normal
    // Data that is typical (expected) and should be accepted by the system.
    { assert: JSON.stringify(removeDups({ data: 'a', next: { data: 'b', next: { data: 'a', next: null } } })), expected: JSON.stringify({ data: 'a', next: { data: 'b', next: null } }) },
    { assert: JSON.stringify(removeDups({ data: 'a', next: { data: 'b', next: { data: 'c', next: { data: 'd', next: null } } } })), expected: JSON.stringify({ data: 'a', next: { data: 'b', next: { data: 'c', next: { data: 'd', next: null } } } }) },

    // Boundary data (extreme data, edge case)
    // Data at the upper or lower limits of expectations that should be accepted by the system.
    { assert: JSON.stringify(removeDups({ data: 'a', next: null })), expected: JSON.stringify({ data: 'a', next: null }) },

    // Abnormal data (erroneous data)
    // Data that falls outside of what is acceptable and should be rejected by the system.
    { assert: removeDups(), expected: false },
    { assert: removeDups(0), expected: false },
    { assert: removeDups(''), expected: false },
    { assert: removeDups([]), expected: false },
    { assert: removeDups({}), expected: false },
    { assert: removeDups(NaN), expected: false },
    { assert: removeDups(null), expected: false },
    { assert: removeDups(false), expected: false },
    { assert: removeDups(new Set()), expected: false },
    { assert: removeDups(new Map()), expected: false },
    { assert: removeDups(undefined), expected: false }
];

// Run tests
testCases.forEach((test, index) => {
    const currentTest = `# Test ${index + 1}`;

    console.group(currentTest);
    console.log('Actual:', test.assert);
    console.log('Expected:', test.expected);
    console.log(test.assert === test.expected ? '🤘 Test PASSED 🤘' : '👎 Test FAILED 👎', '\n');
    console.groupEnd(currentTest);
});

File: Linked Lists/removeDups.py
def removeDups(linkedList={}):
    '''
    Solution 1 - Set track

    Complexity Analysis
    O(n) time | O(n) space

    Remove duplicates from linked list

    dict: linkedList 
    return: Linked list with no duplicates
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedList, dict) or bool(linkedList) == False):
        print('Arguments should be a valid non-empty dictionary')
        return False

    trackedData = set()

    prevNode = dict()
    currNode = linkedList # head

    # Traverse linked list till its tail
    while (currNode != None):
        if (currNode['data'] in trackedData):
            prevNode['next'] = currNode['next']
        else:
            trackedData.add(currNode['data'])
            prevNode = currNode

        currNode = currNode['next']

    return linkedList

def removeDups(linkedList={}):
    '''
    Solution 2 - Inner loop

    Complexity Analysis
    O(n^2) time | O(1) space

    Remove duplicates from linked list

    dict: linkedList 
    return: Linked list with no duplicates
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedList, dict) or bool(linkedList) == False):
        print('Arguments should be a valid non-empty dictionary')
        return False

    currNode = linkedList # head

    # Traverse linked list till its tail
    while (currNode != None):
        prevInnerNode = currNode
        innerNode = currNode['next']

        while (innerNode != None):
            if (currNode['data'] == innerNode['data']):
                prevInnerNode['next'] = innerNode['next']

            prevInnerNode = innerNode
            innerNode = innerNode['next']

        currNode = currNode['next']

    return linkedList

# Test cases (black box - unit testing)
testCases = [
    # Normal
    # Data that is typical (expected) and should be accepted by the system.
    { 'assert': removeDups({ 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'a', 'next': None } } }), 'expected': { 'data': 'a', 'next': { 'data': 'b', 'next': None } } },
    { 'assert': removeDups({ 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'c', 'next': { 'data': 'd', 'next': None } } } }), 'expected': { 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'c', 'next': { 'data': 'd', 'next': None } } } } },
    { 'assert': removeDups({ 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'c', 'next': { 'data': 'c', 'next': { 'data': 'd', 'next': { 'data': 'b', 'next': None } } } } } }), 'expected': { 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'c', 'next': { 'data': 'd', 'next': None } } } } },

    # Boundary data (extreme data, edge case)
    # Data at the upper or lower limits of expectations that should be accepted by the system.
    { 'assert': removeDups({ 'data': 'a', 'next': None }), 'expected': { 'data': 'a', 'next': None } },

    # Abnormal data (erroneous data)
    # Data that falls outside of what is acceptable and should be rejected by the system.
    { 'assert': removeDups(), 'expected': False },
    { 'assert': removeDups(0), 'expected': False },
    { 'assert': removeDups(''), 'expected': False },
    { 'assert': removeDups([]), 'expected': False },
    { 'assert': removeDups(()), 'expected': False },
    { 'assert': removeDups({}), 'expected': False },
    { 'assert': removeDups(None), 'expected': False },
    { 'assert': removeDups(False), 'expected': False }
]

# Run tests
for (index, test) in enumerate(testCases):
    print(f'# Test {index + 1}')
    print(f'Actual: {test["assert"]}')
    print(f'expected: {test["expected"]}')
    print('🤘 Test PASSED 🤘' if test["assert"] == test["expected"] else '👎 Test FAILED 👎', '\n')


